src/Npgsql/NpgsqlTransaction.cs

```diff
@@ -393,6 +421,32 @@ void CheckReady()
                 throw new InvalidOperationException("This NpgsqlTransaction has completed; it is no longer usable.");
         }
 
+        static bool RequiresQuoting(string identifier)
+        {
+            var first = identifier[0];
+            if (!char.IsLower(first) && first != '_')
+                return true;
+
+            for (var i = 1; i < identifier.Length; i++)
+            {
+                var c = identifier[i];
+
+                if (char.IsLower(c) || char.IsDigit(c))
+                    continue;
+
+                switch (c)
+                {
+                case '_':
+                case '$':  // yes it's true
+                    continue;
+                }
+
+                return true;
+            }
+
+            return false;
+        }
```

Comment:
By the way `IsDigit` returns `true` not only for Arabic numbers (0, 1, etc.), but for any number. Is it okay?

```suggestion
        static bool RequiresQuoting(string identifier)
        {
            Debug.Assert(identifier.Length > 0);
            
            var first = identifier[0];
            if (first != '_' && !char.IsLower(first))
                return true;

            foreach (var c in identifier.AsSpan(1))
                if (c != '_' && c != '$' && !char.IsLower(c) && !char.IsDigit(c))
                    return true;

            return false;
        }
```